src/swarm/neo/util/VoidBufferAsArrayOf.d

```diff
@@ -0,0 +1,159 @@
+    has been observed to cause segfaults, e.g.:
+
+    ---
+        void[]* void_array; // acquired from somewhere
+
+        struct S { int i; hash_t h; }
+        auto s_array = cast(S[]*)void_array;
+        s_array.length = 23;
+    ---
+
+    The exact reason for the segfaults is not known, but this usage appears to
+    lead to corruption of internal GC data (possibly type metadata associated
+    with the array's pointer).
+
+    Sizing the array first, then casting is fine, e.g.:
+
+    ---
+        void[]* void_array; // acquired from somewhere
+
+        struct S { int i; hash_t h; }
+        (*void_array).length = 23 * S.sizeof;
+        auto s_array = cast(S[])*void_array;
+    ---
+
+    The helper VoidBufferAsArrayOf simplifies this procedure and removes the
+    risk of undefined behaviour by always handling the void[] as a void[]
+    internally.
+
+    Copyright: Copyright (c) 2017 sociomantic labs GmbH. All rights reserved
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module swarm.neo.util.VoidBufferAsArrayOf;
+
+import ocean.transition;
+
+/*******************************************************************************
+
+    Struct template which wraps a void[] with an API allowing it to be safely
+    used as an array of another type.
+
+    Params:
+        T = element type of array which the API mimics
+
+*******************************************************************************/
+
+public struct VoidBufferAsArrayOf ( T )
+{
+    static assert(!is(T == void));
+
+    private void[]* buffer;
+
+    invariant ( )
+    {
+        assert(this.buffer.length % T.sizeof == 0);
+    }
+
+    public size_t length ( )
+    {
+        return this.buffer.length / T.sizeof;
+    }
+
+    public void length ( size_t len )
+    {
+        this.buffer.length = len * T.sizeof;
+        enableStomping(*this.buffer);
+    }
+
+    public void opCatAssign ( T elem )
+    {
+        this.length = this.length + 1;
+        this.opIndexAssign(elem, this.length - 1);
+    }
+
+    public T* opIndex ( size_t t_index )
+    in
+    {
+        assert(t_index < this.length);
+    }
+    body
+    {
+        return cast(T*)(this.buffer.ptr + (t_index * T.sizeof));
+    }
+
+    public void opIndexAssign ( T elem, size_t t_index )
+    {
+        *this.opIndex(t_index) = elem;
+    }
+
+    public T[] opSlice ( )
+    {
+        return (cast(T*)this.buffer.ptr)[0..this.length];
+    }
+
+    public T[] opSlice ( size_t start, size_t end )
+    {
+        return (cast(T*)this.buffer.ptr)[start..end];
+    }
+
+    public int opApply ( int delegate ( ref T ) dg )
+    {
+        int res;
+        for ( size_t i = 0; i < this.length; i++ )
+        {
+            res = dg(*this.opIndex(i));
+            if ( res )
+                break;
+        }
+        return res;
```

Comment:
Suggest
```d
foreach (ref element; cast(T[])*this.buffer)
    if (int stop = dg(element))
        return stop;
return 0;
```

Reply:
_That's not the opApply recipe!!!_

(Yeah.)